lib/cosmos/interfaces/interface.rb

```diff
@@ -180,10 +180,20 @@ def write_interface
     def read
       raise "Interface not connected for read: #{@name}" unless connected? && read_allowed?
 
+      first = true
       loop do
-        # Read data for a packet
-        data = read_interface()
-        return nil unless data
+        # Protocols may have cached data for a packet, so initially just inject a blank string
+        # Otherwise we can hold off outputing other packets where all the data has already
+        # been received
+        if !first or @read_protocols.length <= 0
```

Comment:
This seems backwards to me. Don't you want to first read_interface to get data and then let the protocols figure out the data? What do you initially send the protocols a blank string of data and then call read_interface every time after?